Generate Java docs through docs_generate.zig

eatonphil opened the pull request:
This PR:

* Adds Java docs generated through docs_generate.zig. View java/README.md for the output.
* Modifies the Node, Go and Java client validation passes in docs_generate.zig to target the current commit. Not what is deployed in the package manager.
* Makes links outside of the client README.mds absolute so they will work when these client READMEs are copied into docs.tigerbeetle.com.

## Pre-merge checklist

* [x] I am very sure this PR could not affect performance.

batiati commented on src/clients/java/docs.zig:
```diff
@@ -0,0 +1,399 @@
+    .create_accounts_example = 
+    \\AccountBatch accounts = new AccountBatch(1);
+    \\accounts.add();
+    \\accounts.setId(UInt128.asBytes(new java.math.BigInteger("137")));
```
That's really a good example: I think it's not the way we should encourage using 128-bit IDs in Java, because we are heap allocating three objects (and because it's so verbose!).

The intended usage of `BigInteger` in our API is for handling numeric literals that cannot be expressed as a Java signed 64-bit long, like this:

In Zig:
```zig
assert(@bitCast(u128, [2]i64{ -100, -1000 }) == 340282366920938445035077277795926146972);
```

In Java:
```java
final var value = UInt128.asBigInteger(-100, -1000);
final var expected = new BigInteger("340282366920938445035077277795926146972");
assertEquals(expected, value);
```

**Suggestions: **

1. As long `"137"` is can be expressed as a Java `long` we don't need a `BigInteger` for that, so we can just do `accounts.setId(UInt128.asBytes(137))`, although we're still allocating a byte array.

2. We can express a 128-bit integer using two `long`, so `accounts.setId(137, 0)` is the preferable usage here, no heap allocation is made. I agree that this second argument `0` can be somehow confusing, especially in an introductory example.

3. We can introduce a new overload for all 128-bit setters expecting only one `long` argument, making it as simple as `accounts.setId(137)`.

eatonphil replied:
Good suggestions! I've added 3 to java_bindings.zig. I'll swap the examples over.

batiati commented on src/clients/java/README.md:
```diff
@@ -1,184 +1,530 @@
+## Setup
 
-## Examples
+First, create `pom.xml` and copy this into it:
```
You deserve a medal 🥇🏅🏵️

batiati reviewed (commented):
Amazing work @eatonphil

batiati commented on src/clients/java/README.md:
```diff
@@ -1,184 +1,530 @@
+Run:
 
-A client needs to be configured with a `clusterID` and `replicaAddresses`. 
-The `Client` class is thread-safe and for better performance, a single instance should be shared between multiple concurrent tasks.
```
```
The `Client` class is thread-safe and for better performance, a single instance should be shared between multiple concurrent tasks.

Multiple clients can be instantiated in case of connecting to more than one TigerBeetle cluster.
```

That is the only piece of information I missed in the new README. I think we should insert a similar note for all clients (changing only language-specific terms, like class, instance, etc.)

eatonphil replied:
Good point but I don't think the Node client is actually thread safe. Is the Go client?

batiati replied:
Actually, the Node client is *kind of* thread-safe, and it does not support creating multiple instances (the node module has a static context). It's on my TODO list to refactor this client using the tb_client.

All other clients share the same principle, thread-safe, and single-instance for each TigerBeetle cluster.

eatonphil replied:
kk, updating with your suggested two sentences.

batiati commented on src/clients/java/docs.zig:
```diff
@@ -145,7 +145,12 @@ pub const JavaDocs = Docs{
+            \\The 128-bit fields like `id` and `user_data` have a few
+            \\overrides to make it easier to integrate. You can either
+            \\pass in a long, a pair of longs (least and most
+            \\significant bits), or a `UInt128`.
```
```suggestion
            \\significant bits), or a `byte[]`.
```
The overloaded setter accepts an array of 16 bytes `byte[]`, likewise, all getters can return an array too.

`UInt128` isn't an instantiable class, it's a helper with static methods and enums for converting 128-bit little-endian unsigned integers between instances of `long`, `UUID`, `BigInteger` and `byte[]`.

For example:
```java
// Using long
account.setId(137);
long id = account.getId(UInt128.LeastSignificant);
assertEquals(137L, id);
assertEquals(0L, account.getId(UInt128.MostSignificant));
```

```java
// Using bytes
byte[] bytes = new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6 };
account.setId(bytes);
byte[] id = account.getId();
assertArrayEquals(bytes, id);
```

eatonphil replied:
Edited with your notes, thank you!

eatonphil commented on src/clients/java/java_bindings.zig:
```diff
@@ -680,6 +680,42 @@ fn emit_u128_batch_accessors(
+    , .{
+        .visibility = if (is_private or is_read_only) @as([]const u8, "") else "public ",
+        .property = to_case(field.name, .pascal),
+    });
```
Sorry for the diff here @batiati it got run through zig fmt.

eatonphil committed:
Adds Java docs generated through docs_generate.zig. View
java/README.md for the output.

Modifies the Node, Go and Java client validation passes in
docs_generate.zig to target the current commit. Not what is deployed
in the package manager.

Makes links outside of the client README.mds absolute so they will
work when these client READMEs are copied into docs.tigerbeetle.com.

cb22 commented on src/clients/node/README.md:
```diff
@@ -7,39 +7,47 @@ The TigerBeetle client for Node.js.
-* [Fedora](./scripts/test_install_on_fedora.sh)
-* [Ubuntu](./scripts/test_install_on_ubuntu.sh)
-* [RHEL](./scripts/test_install_on_rhelubi.sh)
+* [Alpine](https://github.com/tigerbeetledb/tigerbeetle/blob/main/src/clients/node/scripts/scripts/test_install_on_alpine.sh)
```
Looks like a double `scripts/scripts` on these

cb22 replied:
(well, this comment should be down below :D)

eatonphil replied:
hehe fixed thanks!

batiati reviewed (approved):
LGTM (looks great!!)

eatonphil commented:
Thanks for the review @batiati and @cb22 !
